src/Npgsql/NpgsqlConnector.cs

```diff
@@ -1548,20 +1529,51 @@ void DoCancelRequest(int backendProcessId, int backendSecretKey)
+            UserCancellationToken = cancellationToken;
+            ReadBuffer.Cts.ResetCts();
+
+            AttemptPostgresCancellation = attemptPgCancellation;
+            return _cancellationTokenRegistration =
+                cancellationToken.Register(static c => ((NpgsqlConnector)c!).PerformUserCancellation(), this);
+        }
+
         /// <summary>
-        /// Resets cancellation-related state to prepare for a new cancellable operation.
+        /// Starts a new cancellable operation within an ongoing user action. This should only be used if a single user
+        /// action spans several different actions which each has its own cancellation tokens. For example, a command
+        /// execution is a single user action, but spans ExecuteReaderQuery, NextResult, Read and so forth.
         /// </summary>
-        /// <param name="userCancellationToken">
-        /// The cancellation token provided by the user for the operation. This is stored on the connector, and will be referenced
-        /// by the <see cref="OperationCanceledException"/> if one is thrown.
+        /// <remarks>
+        /// Only one level of nested operations is supported. It is an error to call this method if it has previously
+        /// been called, and the returned <see cref="CancellationTokenRegistration"/> was not disposed.
+        /// </remarks>
+        /// <param name="cancellationToken">
+        /// The cancellation token provided by the user. Callbacks will be registered on this token for executing the
+        /// cancellation, and the token will be included in any thrown <see cref="OperationCanceledException"/>.
+        /// </param>
+        /// <param name="attemptPgCancellation">
+        /// If <see langword="true" />, PostgreSQL cancellation will be attempted when the user requests cancellation or
+        /// a timeout occurs, followed by a client-side socket cancellation once
+        /// <see cref="NpgsqlConnectionStringBuilder.CancellationTimeout"/> has elapsed. If <see langword="false" />,
+        /// PostgreSQL cancellation will be skipped and client-socket cancellation will occur immediately.
         /// </param>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        internal void ResetCancellation(CancellationToken userCancellationToken)
+        internal CancellationTokenRegistration StartNestedCancellableOperation(
+            CancellationToken cancellationToken = default,
+            bool attemptPgCancellation = true)
         {
-            _cancellationRequested = false;
-            _userCancellationRequested = false;
-            UserCancellationToken = userCancellationToken;
-            ReadBuffer.Cts.ResetCts();
+            cancellationToken.ThrowIfCancellationRequested();
+
+            UserCancellationToken = cancellationToken;
+            AttemptPostgresCancellation = attemptPgCancellation;
+
+            return _cancellationTokenRegistration =
```

Comment:
You're still not disposing the previous token, don't you?

Reply:
You mean the previous registration, right? If so, the idea is that it's the caller responsibility to always do that. If you call StartNestedCancellableOperation, you must dispose it the returned registration before StartNestedCancellableOperation gets called again. Sounds ok?

Reply:
I'm talking about outer scope's registration. If you start the user action with a token, and then nested, you'll lose the previous registration.

Reply:
So here are the assumptions/constraints around this:
* If user action is started with a cancellation token, it will never have a nested scope. This corresponds to the simple case, where there's only one operation/function call which is represented by the user action. In other words, if there's a nested scope (complex case), the containing user action above cannot have a cancellation token.
* The two examples of complex cases are command execution and COPY, where the user action spans multiple user function calls, which can have different cancellation tokens. In those cases, the user action should never receive a user cancellation token - only the nested operations within it.
* There can never be a nested operation within a nested operation (we've already discussed this).

I agree that this isn't obvious from the code, and we could make some API changes to make this explicit (and/or add debugging assertions to make sure the above invariants aren't violated). But we can do that later, assuming that there's no bugs.

Reply:
Okay